### PR TITLE
DISABLE_TALKIE

### PR DESCRIPTION
**Pull request by fdarosa2663**
Save 2336 bytes by replacing talkie functions with beeper, will create supporting POD to document

### CODE REVIEW

File: sound/sound_library.h [hunk 1]
```diff
@@ -8,6 +8,7 @@ enum SayType {
   SAY_WHOLE,
 };
 
+#ifdef SAY_COLOR_LIST
```
Comment (fdarosa2663) [inline sound/sound_library.h:11]: if sound_library is used and we don't use SAY_COLOR_LIST save memory for EFFECT and enum

File: styles/length_finder.h [hunk 2]
```diff
@@ -34,7 +34,7 @@ class LengthFinder {
     if (say_it_ && millis() - last_speak_ > 3000) {
       last_speak_ = millis();
       say_it_ = false;
-#ifdef ENABLE_AUDIO      
+#if defined(ENABLE_AUDIO) && !defined(DISABLE_TALKIE)
```
Comment (profezzorn) [inline styles/length_finder.h:37]: This should be an error.
Lengthinder is useless without talkie.

File: sound/sound_library.h [hunk 3]
```diff
@@ -265,9 +267,11 @@ class SoundLibrary {
   void SaySettingsMenu() { Play("msetsub.wav"); }
   void SayStyleSettings() { Play("stylstm.wav"); }
 
+#ifdef SAY_COLOR_LIST
```
Comment (profezzorn) [inline sound/sound_library.h:270]: This seems unrelated

File: props/prop_base.h [hunk 4]
```diff
@@ -114,6 +114,10 @@ class SoundQueue {
 };
 #endif
 
+#ifdef DISABLE_TALKIE
+#include "../sound/sound_library.h"
```
Comment (profezzorn) [inline props/prop_base.h:118]: Is this used anywhere?

File: props/prop_base.h [hunk 5]
```diff
@@ -603,9 +607,16 @@ class PropBase : CommandParser, Looper, protected SaberBase {
   bad_blade:
     STDOUT.println("BAD BLADE");
 #ifdef ENABLE_AUDIO
+#ifdef DISABLE_TALKIE
```
Comment (profezzorn) [inline props/prop_base.h:610]: These needs to be moved to a common location, and have comments that describe what they sound like.

File: props/saber_fett263_buttons.h [hunk 6]
```diff
@@ -685,6 +685,10 @@ CUSTOM SOUNDS SUPPORTED (add to font to enable):
 #define FETT263_USE_SETTINGS_MENU
 #endif
 
+#if defined(FETT263_SAY_COLOR_LIST) || defined(FETT263_SAY_COLOR_LIST_CC)
+#define SAY_COLOR_LIST
```
Comment (profezzorn) [inline props/saber_fett263_buttons.h:689]: unrelated?

**Review by profezzorn — COMMENTED** [review]
So, it seems that this change doesn't actually disable the talkie class itself.
I suspect that much more memory would be saved if the talkie class was completely left out.

File: styles/length_finder.h (hunk 2 above)
Comment (fdarosa2663) [reply styles/length_finder.h:37]: thought user could just use Serial Monitor in this case?

**Comment by fdarosa2663** [comment]
I'll try removing talkie class completely and other suggestions and make updates

File: props/saber_fett263_buttons.h (hunk 6 above)
Comment (fdarosa2663) [reply props/saber_fett263_buttons.h:689]: it's an additional memory saving for sound_library when SAY_COLOR_LIST defines are not enabled.  So tangentially related?

File: props/prop_base.h (hunk 5 above)
Comment (fdarosa2663) [reply props/prop_base.h:606]: I assume you mean a comment explaining what the define does and the beeps that replace the messages?  Where would you recommend?

I had attempted to centralize all errors to a single file but it would not compile so I left the change to beeper in the existing location.

**Comment by fdarosa2663** [comment]
Looks like completely removing Talkie class frees up 7576 bytes, so definitely better.

Should I go ahead and update the other props that are currently using talkie, as the define will now create a compile error when used - or - should I just notify them and leave it to those creators to handle?

Just need a few replies to above conversations to finish up.

**Comment by profezzorn** [comment]
7.5k, not bad :)

I think it depends on the props, and how complicated they are and if they are actively maintained or not.
I think it's ok to have compile errors if you try to use a feature that needs talkie and also have DISABLE_TALKIE in your config file. We just need to warn people that such things can happen in the documentation for the define.  (LengthFinder would be a good example of this IMHO.)

For the most part, I think it's fine to leave the props as is and just let them cause errors with DISABLE_TALKIE, but if you want to make things better for users, you can:

1. Find cases where talkie could be replaced with beeps, and add code to the props for that.
2. Find cases where talkie is entirely optional, and just use #ifndef to make it not do that.
3. Find props where talkie is integral, and removing it will make no sense and add #error messages for them.

I would probably do all of this in a followup PR, rather than trying to include any of it in this PR though.

We may also be able to create a dummy talkie class which causes more readable errors if you try to call Say() on it, although I would have to read up on how to do that.

Looks like this is how you create custom error messages for method calls:
https://stackoverflow.com/questions/34745581/forbids-functions-with-static-assert/34746077#34746077

**Comment by fdarosa2663** [comment]
OK, for now I'll leave the other props to handle how they want to.  I plan to do a write up on POD for the define with the replacement beeps for the respective talkie we replace.

I probably need help on the static assert, unfortunately I didn't really follow all of that.  I suppose I could just take the define our of lengthfinder and let the compile error handle if a user tried to use.

For a comment on the function where would you recommend it?  In the .ino?

File: sound/talkie.h [hunk 7]
```diff
@@ -1,6 +1,13 @@
 #ifndef SOUND_TALKIE_H
 #define SOUND_TALKIE_H
 
+/* DISABLE_TALKIE define will save memory by replacing the default
```
Comment (fdarosa2663) [inline sound/talkie.h:4]: Added comment to talkie.h for define, let me know if you would prefer somewhere else.

**Comment by fdarosa2663** [comment]
Made updates, added comment for define in talkie.h but can move if you have a better location.

**Comment by profezzorn** [comment]
I did a chunk of this work, not sure if it makes sense to continue this PR or if it's easier to create a new one.

**Comment by fdarosa2663** [comment]
We can close this one, do you need me to do anything from my end, I should have time in the morning?

**Comment by profezzorn** [comment]
Error.h still needs to be updated or DISABLE_TALKIE isn't going to actually work.
